### ISSUE TITLE
OpenMP Target Missing support for non-trivial reduction types on GPUs

### ISSUE THREAD

**Issue by martagarciamartinez** [llvm/llvm-project#56632]
While building Kokkos with TESTS and EXAMPLES ON on V100 I found this error:

==============================================
[ 29%] Building CXX object core/unit_test/CMakeFiles/KokkosCore_UnitTest_OpenMPTarget.dir/openmptarget/TestOpenMPTarget_Reducers_d.cpp.o
clang-15: /home/yeluo/.jacamar-ci/builds/a5MjzozQ/001/intel-dga/automatedtests/yeluo/test_llvm/llvm-project/clang/lib/CodeGen/CGExprAgg.cpp:2067: void clang::CodeGen::CodeGenFunction::EmitAggregateCopy(clang::CodeGen::LValue, clang::CodeGen::LValue, clang::QualType, clang::CodeGen::AggValueSlot::Overlap_t, bool): Assertion `(Record->hasTrivialCopyConstructor() || Record->hasTrivialCopyAssignment() || Record->hasTrivialMoveConstructor() || Record->hasTrivialMoveAssignment() || Record->hasAttr<TrivialABIAttr>() || Record->isUnion()) && "Trying to aggregate-copy a type without a trivial copy/move " "constructor or assignment operator"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
==============================================

Based on a discussion with Johannes during his LLVM Office Hours, it seems that the problem is that 
LLVM does not support non trivial copy-constructors in device reductions => LLVM Team would need to find someone who could implement support for non trivial GPU reductions.

I have files and reproducer. Please feel free to contact me.

**Comment by ye-luo** [comment]
Is it because of missing user-defined reduction?

**Comment by jdoerfert** [comment]
No, we don't support non trivially copyable types in device reductions. The current impl which we want to replace really badly is using memcpy and not copy constructors. We should obviously not crash the compilation like that but we won't support it properly unless we reimplement device reduction.
